input/kafka_avro.go

```diff
@@ -0,0 +1,137 @@
+package input
+
+import (
+	"context"
+	"encoding/binary"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/housepower/clickhouse_sinker/config"
+	"github.com/housepower/clickhouse_sinker/model"
+	"github.com/housepower/clickhouse_sinker/statistics"
+	"github.com/housepower/clickhouse_sinker/util"
+	"github.com/pkg/errors"
+	"github.com/riferrei/srclient"
+	"go.uber.org/zap"
+	"io"
+)
+
+var _ Inputer = (*KafkaAvro)(nil)
+
+// KafkaAvro implements input.Inputer
+type KafkaAvro struct {
+	consumer             *kafka.Consumer
+	cfg                  *config.Config
+	taskCfg              *config.TaskConfig
+	stopped              chan struct{}
+	putFn                func(msg model.InputMessage)
+	schemaRegistryClient *srclient.SchemaRegistryClient
+}
+
+// NewKafkaAvro get instance of kafka reader
+func NewKafkaAvro() *KafkaAvro {
+	return &KafkaAvro{}
+}
+
+// Init Initialise the kafka instance with configuration
+func (k *KafkaAvro) Init(cfg *config.Config, taskCfg *config.TaskConfig, putFn func(msg model.InputMessage), cleanupFn func()) (err error) {
+	k.cfg = cfg
+	k.taskCfg = taskCfg
+	k.stopped = make(chan struct{})
+	k.putFn = putFn
+	kfkCfg := &cfg.Kafka
+
+	offset := kafka.OffsetEnd
+	if k.taskCfg.Earliest {
+		offset = kafka.OffsetBeginning
+	}
+
+	k.consumer, err = kafka.NewConsumer(&kafka.ConfigMap{
+		"bootstrap.servers":  kfkCfg.Brokers,
+		"security.protocol":  kfkCfg.Security["protocol"],
+		"sasl.mechanism":     kfkCfg.Sasl.Mechanism,
+		"sasl.username":      kfkCfg.Sasl.Username,
+		"sasl.password":      kfkCfg.Sasl.Password,
+		"group.id":           k.taskCfg.ConsumerGroup,
+		"session.timeout.ms": 6000,
+		"auto.offset.reset":  offset,
+	})
+	k.schemaRegistryClient = srclient.CreateSchemaRegistryClient(kfkCfg.SchemaRegistryURL)
+	return k.consumer.SubscribeTopics([]string{k.taskCfg.Topic}, nil)
+}
+
+// kafka main loop
+func (k *KafkaAvro) Run(_ context.Context) {
+LOOP_ARMO:
+	for {
+
+		msg, err := k.consumer.ReadMessage(-1)
+
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				util.Logger.Info("KafkaAvro.Run quit due to context has been canceled", zap.String("task", k.taskCfg.Name))
+				break LOOP_ARMO
+			} else if errors.Is(err, io.EOF) {
+				util.Logger.Info("KafkaAvro.Run quit due to reader has been closed", zap.String("task", k.taskCfg.Name))
+				break LOOP_ARMO
+			} else {
+				statistics.ConsumeMsgsErrorTotal.WithLabelValues(k.taskCfg.Name).Inc()
+				err = errors.Wrap(err, "")
+				util.Logger.Error("kafka.Reader.FetchMessage failed", zap.String("task", k.taskCfg.Name), zap.Error(err))
+				continue
+			}
+		}
+
+		if len(msg.Value) < 5 {
+			continue
```

Comment:
Need to invoke putFn to pass this message to the parser phase even if the message value is invalid. The parser will mark it as parsing error and not blocking next messages processing.